utils/char16trie/src/char16trie.rs

```diff
@@ -0,0 +1,396 @@
+        }
+    }
+
+    fn next_impl(&mut self, pos: usize, in_unit: i32) -> TrieResult {
+        let mut node = self.get(pos);
+        let mut pos = pos + 1;
+        loop {
+            if node < MIN_LINEAR_MATCH {
+                return self.branch_next(pos, node as usize, in_unit);
+            } else if node < MIN_VALUE_LEAD {
+                // Match the first of length+1 units.
+                let length = node - MIN_LINEAR_MATCH;
+                if in_unit == self.get(pos).into() {
+                    pos += 1;
+                    if length == 0 {
+                        self.remaining_match_length = None;
+                        self.pos = Some(pos);
+                        node = self.get(pos);
+                        if node >= MIN_VALUE_LEAD {
+                            return self.value_result(pos);
+                        }
+                        return TrieResult::NoValue;
+                    }
+                    self.remaining_match_length = Some(length as usize - 1);
+                    self.pos = Some(pos);
+                    return TrieResult::NoValue;
+                }
+                // No match
+                break;
+            } else if (node & VALUE_IS_FINAL) != 0 {
+                // No further matching units.
+                break;
+            } else {
+                // Skip intermediate value.
+                pos = skip_node_value(pos, node);
+                node &= NODE_TYPE_MASK;
+            }
+        }
+        self.stop();
+        TrieResult::NoMatch
+    }
+
+    fn stop(&mut self) {
+        self.pos = None;
+    }
+
+    fn jump_by_delta(&self, pos: usize) -> usize {
+        let delta = self.get(pos);
+        if delta < MIN_TWO_UNIT_DELTA_LEAD {
+            // nothing to do
+            pos + 1 + delta as usize
+        } else if delta == THREE_UNIT_DELTA_LEAD {
+            let delta = ((self.get(pos + 1) as usize) << 16) | (self.get(pos + 2) as usize);
+            pos + delta + 3
+        } else {
+            let delta =
+                ((delta - MIN_TWO_UNIT_DELTA_LEAD) as usize) << 16 | (self.get(pos + 1) as usize);
+            pos + delta + 2
+        }
+    }
+
+    fn skip_value(&self, pos: usize) -> usize {
+        let lead_byte = self.get(pos);
+        skip_value(pos + 1, lead_byte & 0x7fff)
+    }
+
+    fn skip_delta(&self, pos: usize) -> usize {
+        let delta = self.get(pos);
+        if delta < MIN_TWO_UNIT_DELTA_LEAD {
+            pos + 1
+        } else if delta == THREE_UNIT_DELTA_LEAD {
+            pos + 3
+        } else {
+            pos + 2
+        }
+    }
+
+    fn value_result(&self, pos: usize) -> TrieResult {
+        let node = self.get(pos) & VALUE_IS_FINAL;
+        let value = self.get_value(pos);
+        match node {
+            VALUE_IS_FINAL => TrieResult::FinalValue(value),
+            _ => TrieResult::Intermediate(value),
+        }
+    }
+
+    pub fn get_value(&self, pos: usize) -> i32 {
```

Comment:
*If* you always call this whenever you return a TrieResult that has a value, then the pos is always after the value, and then this public function, when called explicitly, will never return a real value, right? Making it useless to have this as public API?

My preference is for this code to work like in C++ & Java, and have the user call get_value() only when TrieResult.has_value() *and the user cares to see the value*.

Reply:
I agree this should not be public if we're using the Result the way it is defined above. Let's iterate on that conversation before I change anything here.

Reply:
I also prefer the TrieResult-based API.